includes/data-port/class-sensei-data-port-job

```diff
@@ -0,0 +1,347 @@
+	 *
+	 * @return array The logs.
+	 */
+	public function get_logs( $offset = 0 , $limit = 0 ) {
+		if ( $limit > 0 ) {
+			return array_slice( $this->logs, $offset, $limit );
+		} else {
+			return $this->logs;
+		}
+	}
+
+	/**
+	 * Delete any stored state for this job.
+	 */
+	public function clean_up() {
+		delete_option( self::get_option_name( $this->job_id ) );
+	}
+
+	/**
+	 * Get the completion status of the job.
+	 *
+	 * @return array
+	 */
+	public function get_status() {
+		if ( $this->is_completed ) {
+			return [
+				'status' => 'completed',
+				'percentage' => 100,
+			];
+		}
+
+		return [
+			'status' => 'running',
+			'percentage' => $this->percentage,
+		];
+	}
+
+	/**
+	 * Creates a new data port job.
+	 *
+	 * @param string $job_id    Unique id of the job.
+	 * @param array $sub_jobs   Sub-job levels.
+	 *
+	 * @return Sensei_Data_Port_Job The newly created job.
+	 */
+	abstract public static function create( $job_id, $sub_jobs );
+
+	/**
+	 * Do question porting.
+	 *
+	 * @param array $question_state  Question sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                               'completed' when the port is completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_question_port( &$question_state );
+
+	/**
+	 * Do lesson porting.
+	 *
+	 * @param array $lesson_state  Lesson sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                             'completed' when the port is completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_lesson_port( &$lesson_state );
+
+	/**
+	 * Do course porting.
+	 *
+	 * @param array $course_state  Course sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                             'completed' when the port id completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_course_port( &$course_state );
+
+	/**
+	 * Serialize state to JSON.
+	 *
+	 * @return array
+	 */
+	public function jsonSerialize() {
+		return [
+			's' => $this->sub_jobs_state,
+			'l' => $this->logs,
+			'r' => $this->results,
+			'c' => $this->is_completed,
+			'p' => $this->percentage,
+		];
+	}
+
+	/**
+	 * Restore state from JSON.
+	 */
+	private function restore_from_json( $json_string ) {
+		$json_arr = json_decode( $json_string, true );
+
+		if ( ! $json_arr ) {
+			return;
+		}
+
+		$this->sub_jobs_state = $json_arr['s'];
+		$this->logs           = $json_arr['l'];
+		$this->results        = $json_arr['r'];
+		$this->is_completed   = $json_arr['c'];
+		$this->percentage     = $json_arr['p'];
+	}
+
+	/**
+	 * Add an entry to the logs.
+	 *
+	 * @param string $name      Name of the entity this log applies to.
+	 * @param string $message   Log message.
+	 * @param string $id        Id of the entity this log applies to.
+	 */
+	protected function add_log_entry( $name, $message, $id = '' ) {
```

Comment:
We should probably add a log type to specify which level the message applies to.